[MRG+1] FIX: warns when invalid n_components in LinearDiscriminantAnalysis

wdevazelhes opened the pull request:
#### Reference Issues/PRs
Fixes #10048.
Fixes #8956. (The second dimension of scalings will always be thresholded (not only for svd, (see https://github.com/scikit-learn/scikit-learn/issues/8956#issuecomment-376881805)))

#### What does this implement/fix? Explain your changes.
This PR: 
 - Raises a `ChangedBehaviourWarning` when the user sets `n_components` > `min(n_features, n_classes - 1)`. In this case it sets the `max_features` (the number of first components to take) to `min(n_features, n_classes - 1)` (and this way it doesn't take the `n_components` into account anymore). It does not throws an error like PCA not to break user code (cf comment: https://github.com/scikit-learn/scikit-learn/issues/6355#issuecomment-340855091). I should maybe provide a `FutureWarning` and throw an error in the future ? 
- Changes the docstring, saying that we should have `n_components < min(n_features, n_classes - 1)` (and not just `n_components` < `n_classes - 1`)
- Tests that if the condition is verified no warning is thrown otherwise a warning is thrown

I did not check explicitly the dimension (just the presence/absence of warnings) because it can still happen that the dimension is unexpected if input points are colinear. I was thinking to tackle this in another PR (raise a warning in that case and/or return the whole `scalings_` (including zeros) without truncation) (see #11528)

### TODO: 
- [x] Add `FutureWarning`

Commit:
FIX: warns when invalid n_components in LinearDiscriminantAnalysis

amueller commented:
I think future warning sounds good.

amueller commented on sklearn/discriminant_analysis.py:
```diff
@@ -166,8 +167,10 @@ class LinearDiscriminantAnalysis(BaseEstimator, LinearClassifierMixin,
-    n_components : int, optional
-        Number of components (< n_classes - 1) for dimensionality reduction.
+    n_components : int, optional (default=None)
+        Number of components (< min(n_classes - 1, n_features)) for
```
<=, right? = is the default after all...

amueller commented on sklearn/discriminant_analysis.py:
```diff
@@ -443,11 +446,24 @@ def fit(self, X, y):
-                                       self.n_components)
+            if self.n_components > max_components:
+                warnings.warn(
+                    "n_components cannot be superior to min(n_features, "
```
I would say "larger than" not "superior". I'm not a native English speaker and the usage strikes me as odd. Even if it's correct, we have many users that are not native English speakers and will be thrown off by it.

amueller commented on sklearn/discriminant_analysis.py:
```diff
@@ -443,11 +446,24 @@ def fit(self, X, y):
+
         if self.n_components is None:
-            self._max_components = len(self.classes_) - 1
+            self._max_components = max_components
```
I feel the private variable should be called ``_n_components``, not ``_max_components``.

wdevazelhes replied:
I agree, however maybe it could be justified to call it `_max_components` for the case where some inputs of LDA would be collinear ? indeed the `scalings_` would be truncated and have less components than `_n_components` (see issue #11528)

amueller commented on sklearn/tests/test_discriminant_analysis.py:
```diff
@@ -257,6 +260,33 @@ def test_lda_store_covariance():
+@pytest.mark.parametrize('n_features', [3, 5])
+@pytest.mark.parametrize('n_classes', [5, 3])
+def test_lda_dimension_warning(n_classes, n_features):
+    RNG = check_random_state(0)
```
lowercase please. upper case is reserved for module level constants, right? (yes, X violates that, I know).

wdevazelhes replied:
that's right, will do

amueller commented on sklearn/tests/test_discriminant_analysis.py:
```diff
@@ -257,6 +260,33 @@ def test_lda_store_covariance():
+
+    for n_components in [max_components + 1,
+                         max(n_features, n_classes - 1) + 1]:
+        # if n_components < min(n_classes - 1, n_features), raise warning
```
``>`` ?

wdevazelhes replied:
yes, sorry, typo

amueller commented on sklearn/tests/test_discriminant_analysis.py:
```diff
@@ -257,6 +260,33 @@ def test_lda_store_covariance():
+    max_components = min(n_features, n_classes - 1)
+
+    for n_components in [max_components - 1, None, max_components]:
+        # if n_components < min(n_classes - 1, n_features), no warning
```
``<=``?

wdevazelhes replied:
yes, will do

amueller commented on sklearn/tests/test_discriminant_analysis.py:
```diff
@@ -257,6 +260,33 @@ def test_lda_store_covariance():
+        assert_no_warnings(lda.fit, X, y)
+
+    for n_components in [max_components + 1,
+                         max(n_features, n_classes - 1) + 1]:
```
I don't understand the second one?

wdevazelhes replied:
I am not sure about this one indeed, it's just that since I test something just one unit higher than `max_components`, I thought I could test something that is higher than both `n_features` and n_classes - 1 to ensure the test works for any value of `n_components`

GaelVaroquaux replied:
That's good for me. Maybe a small comment that explains this.

wdevazelhes replied:
Done

amueller reviewed (commented):
generally looks good, but the boundary case is unclear in the doc and comments.

Commit:
STY: fix flake8 errors

wdevazelhes commented:
> I think future warning sounds good.

Alright, I'll include one

Commit:
FIX: fixes according to code review https://github.com/scikit-learn/scikit-learn/pull/11526#pullrequestreview-137269580
- fix doc for boundary case using including unequalities
- fix typos
- fix style conventions

Commit:
ENH: add FutureWarning
and fixes test warning assertion

GaelVaroquaux commented on sklearn/discriminant_analysis.py:
```diff
@@ -443,11 +447,30 @@ def fit(self, X, y):
+                    "n_classes - 1) = min(%d, %d - 1) = %d components."
+                    % (X.shape[1], len(self.classes_), max_components),
+                    ChangedBehaviorWarning)
+                future_msg = ("In version 0.22, invalid values for "
```
Maybe it would be useful to say in a few words what invalid means.

GaelVaroquaux reviewed (approved):
Aside from the two minor comments that I made, this is good for me.

Commit:
MAINT: modify warning and add comments regarding review https://github.com/scikit-learn/scikit-learn/pull/11526#pullrequestreview-137513341

Commit:
relaunch appveyor

TomDLT reviewed (approved):
LGTM, but you have a conflict.

You also need to add a whatsnew entry.

Thanks !

Commit:
Merge branch 'master' into fix/warn_n_components_lineardiscriminantanalysis

# Conflicts:
#	sklearn/discriminant_analysis.py
#	sklearn/tests/test_discriminant_analysis.py

Commit:
MAINT: Update FeatureWarning from 0.22 to 0.23

Commit:
DOC: Add what's new entry

Commit:
MAINT: remove useless comment

Commit:
MAINT: remove unused import

Commit:
FIX: add missing import

Commit:
FIX: update FutureWarning in tests too

wdevazelhes commented:
Sorry for the late reply. I resolved the conflict, added a what's new entry, and changed the `FutureWarning` from 0.22 to 0.23 . I guess it should be ready to merge

TomDLT committed:
Update v0.21.rst

TomDLT commented:
Thanks !
